### PR TITLE
allow overriding validator for a field,

### PR DESCRIPTION
**Pull request by cmangeat**
checking that value is not used by another record, given current record id from match dict

do not ovveride validator when one already set

at scaffold side, add validator on reference column and test it

### CODE REVIEW

File: c2cgeoform/schema.py [hunk 1]
```diff
@@ -14,13 +15,30 @@ def deferred_dbsession(node, kw):
     return kw.get('dbsession')
 
 
+def unique_validator(mapper, column, id_column, node, value):
+    dbsession = node.bindings['dbsession']
+    _id = node.bindings['request'].matchdict['id']
+    _id = _id if _id != 'new' else None
+    other_consumer = dbsession.query(mapper).filter(column == value, id_column != _id).first()
```
Comment (arnaud-morvan) [inline c2cgeoform/schema.py]: you do not need the instance, just the count

File: c2cgeoform/schema.py [hunk 2]
```diff
@@ -14,13 +15,30 @@ def deferred_dbsession(node, kw):
...
+    if other_consumer is not None:
+        exc = colander.Invalid(node, '%s use %s.' % (getattr(other_consumer, id_column.name), value), value)
```
Comment (arnaud-morvan) [inline c2cgeoform/schema.py]: I would prefer `_('{} is already used').format(value)`

File: c2cgeoform/schema.py [hunk 3]
```diff
@@ -14,13 +15,30 @@ def deferred_dbsession(node, kw):
...
         super().__init__(*args, **kw)
         self.request = deferred_request
         self.dbsession = deferred_dbsession
 
+    def add_unique_validator(self, column, column_id):
+        validator = partial(unique_validator, self.class_, column, column_id)
+        if self[column.name].validator is None:
+            self[column.name].validator = validator
+        else:
+            self[column.name].validator = colander.All(self[column.name].validator, validator)
```
Comment (arnaud-morvan) [inline c2cgeoform/schema.py]: We do not need to send a request to the database if other validator fails.

File: c2cgeoform/tests/test_schema.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            schema_node.deserialize({'text': 'foo'})
+
+    @patch('c2cgeoform.schema.unique_validator')
+    def test_adding_validator_does_not_overrides_sqlalchemy_ones(self, unique_validator_mock):
+        schema_node = GeoFormSchemaNode(FieldsCollection)
+        schema_node.add_unique_validator(FieldsCollection.text, FieldsCollection.id)
+        with self.assertRaises(colander.Invalid):
+            schema_node.deserialize({'text': 'more than five char'})
+        unique_validator_mock.assert_has_calls(
+            [mock.call(mock.ANY, mock.ANY, mock.ANY, mock.ANY, 'more than five char')])
```
Comment (arnaud-morvan) [inline c2cgeoform/tests/test_schema.py]: could be : 
```
try:
    schema_node.deserialize({'text': 'more than five char'})
    assert False
except colander.Invalid as e:
    assert e.message == "blabla"

File: c2cgeoform/schema.py (hunk 3 above)
Comment (cmangeat) [reply c2cgeoform/schema.py]: Sad, it seems that we have to choose between colander.ALL or colander.ANY, and both of them trigger all validators, expecting all to success, or one not to fail, but even if the first has failed (or succeded for ANY), all validators are triggered.

File: c2cgeoform/tests/test_schema.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+class TestUniqueValidator(unittest.TestCase):
+
+    def test_constraint_on_column_at_sql_alchemy_side_throw_colander_invalid(self):
+        schema_node = GeoFormSchemaNode(FieldsCollection)
+        schema_node.deserialize({'text': 'foo'})
+        try:
+            schema_node.deserialize({'text': 'more than five char'})
+            assert False
+        except colander.Invalid as e:
+            assert 0 == e.children[0].msg.find('Longer than maximum length')
```
Comment (cmangeat) [inline c2cgeoform/tests/test_schema.py]: 'Longer than maximum length' in e.children[0].msg

**Pull request merged**